ddtrace/tracer/abandoned_spans.go

```diff
@@ -17,26 +17,57 @@ import (
 var tickerInterval = time.Minute
 var logSize = 9000
 
-// isBucketNode takes in a list.Element and checks if it is a nonempty
-// list Element
-func isBucketNode(e *list.Element) (*list.List, bool) {
+// castAsBucketNode takes in a list.Element and checks if it is a nonempty
+// list Element. If it can be cast as a list.List, that list will be returned
+// bool flag `true`. If not, it will return `nil` with bool flag `false`
+func castAsBucketNode(e *list.Element) (*list.List, bool) {
+	if e == nil {
+		return nil, false
+	}
 	ls, ok := e.Value.(*list.List)
 	if !ok || ls == nil || ls.Front() == nil {
 		return nil, false
 	}
 	return ls, true
 }
 
-// isSpanNode takes in a list.Element and checks if it is a non-nil
-// span object
-func isSpanNode(e *list.Element) (*span, bool) {
+// castAsSpanNode takes in a list.Element and checks if it is a non-nil
+// span object. If it can be cast as a span, that span will be returned
+// bool flag `true`. If not, it will return `nil` with bool flag `false`
+func castAsSpanNode(e *list.Element) (*span, bool) {
+	if e == nil {
+		return nil, false
+	}
 	s, ok := e.Value.(*span)
 	if !ok || s == nil {
 		return nil, false
 	}
 	return s, true
 }
 
+// findSpanBucket takes in a start time in Unix Nanoseconds and the user
+// configured interval, then finds the bucket that the given span should
+// belong in. All spans within the same bucket should have a start time that
+// is within `interval` nanoseconds of each other.
+func (t *tracer) findSpanBucket(start int64, interval time.Duration) (*list.Element, bool) {
+	for node := t.abandonedSpans.Front(); node != nil; node = node.Next() {
+		bucket, ok := castAsBucketNode(node)
+		if !ok {
+			continue
+		}
+		for spNode := bucket.Front(); spNode != nil; spNode = spNode.Next() {
+			sp, ok := castAsSpanNode(spNode)
+			if !ok {
+				continue
+			}
+			if start >= sp.Start && start-sp.Start < interval.Nanoseconds() {
```

Comment:
hmmm is this different than it was before where it checked if the time was before or after the start by `interval` amount?

Reply:
Yeah, I realized that the previous way might be inaccurate/hard to read. Also, for a newer span comparing to an older bucket, I _think_ the old way might add to the wrong buckets sometimes? Totally correct me if I'm wrong, or if this might need changing.